src/Control/Director.php

```diff
@@ -841,28 +842,16 @@ public static function is_site_url($url)
 
     /**
      * Given a filesystem reference relative to the site root, return the full file-system path.
-     *
-     * @param string $file
-     *
-     * @return string
      */
-    public static function getAbsFile($file)
+    public static function getAbsFile(string $file): string
     {
-        // If already absolute
-        if (Director::is_absolute($file)) {
-            return $file;
-        }
+        $path = Director::getAbsolutePathForFile($file);
 
-        // If path is relative to public folder search there first
-        if (Director::publicDir()) {
-            $path = Path::join(Director::publicFolder(), $file);
-            if (file_exists($path ?? '')) {
-                return $path;
-            }
+        if (!Path::isBasePath(Director::baseFolder(), $path)) {
+            throw new InvalidArgumentException("'$file' must not be outside the project root");
         }
```

Comment:
Added traversal protection here because it seems likely for the return value of `Director::getAbsFile()` to be directly used to read files at the returned path.